### PR TITLE
Move keyword tokenizer to analysis-common module

### PR DESCRIPTION
**Pull request by martijnvg**
Relates to #23658

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 1]
```diff
@@ -548,6 +548,10 @@ private void processNormalizerFactory(
             TokenizerFactory keywordTokenizerFactory,
             Map<String, TokenFilterFactory> tokenFilters,
             Map<String, CharFilterFactory> charFilters) {
+        if (keywordTokenizerFactory == null) {
+            throw new IllegalStateException("keyword tokenizer factory is null");
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: nit: maybe IllegalArgumentException but it really doesn't matter that much here.

File: server/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java [hunk 2]
```diff
@@ -425,12 +437,17 @@ public void testRegisterHunspellDictionary() throws Exception {
 
     // Simple char filter that appends text to the term
     public static class AppendCharFilter extends CharFilter {
-        private final char[] appendMe;
-        private int offsetInAppendMe = -1;
+
+        static Reader append(Reader input, String appendMe) {
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java]: I assume this is an optimization unrelated to the tokenizer move? Can you quickly confirm?

**Review by cbuescher — APPROVED** [review]
LGTM, left one question mostly for my own education.

File: server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: Maybe the message should mention that this requires the `analysis-common` module?

File: server/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java (hunk 2 above)
Comment (nik9000) [reply server/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java]: It isn't so much an optimization as "this should be easier to read". I think.

**Review by nik9000 — APPROVED** [review]
I *wonder* if the keyword field should be in anlaysis-common then? Maybe it is too tied in to core. Either way, I think that this is a good next step and we can handle other things in a followup.

File: server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: @cbuescher I like this keep throwing `IllegalStateException` if we detect this far in starting up that there is no keyword tokenizer then we have detected way to late that we need the analysis-common module and in a sense this is then a misconfiguration, not really a missing setting.

@nik9000 I will update the error message.

**Comment by martijnvg** [comment]
> I wonder if the keyword field should be in anlaysis-common then? Maybe it is too tied in to core. 

@nik9000 This also crossed my mind when making this change (esp. since normalizers need keyword tokenizer to be available in order to work). However it is not part of lucene-core and therefor its rightful place is the analysis-common module, otherwise we can't remove the lucene analysis-common dependency from the server module.

**Pull request merged**